File: datasette_otel_viewer/metrics_math.py
"""Pure metric math shared by the query API and page routes. Mirrors
frontend/src/lib/metricsMath.ts (same test vectors) — keep them in sync."""

from __future__ import annotations

import json
from collections.abc import Iterable

NS = 1_000_000_000


def bucket_start(time_ns: int, step_ns: int) -> int:
    "Floor time_ns to the step grid anchored at the Unix epoch."
    return (time_ns // step_ns) * step_ns


def series_key(
    service_name: str | None, attributes: dict, group_by: list[str] | None
) -> str:
    """Stable series identity. group_by=None keeps the whole attribute set
    (native OTel series identity); a list keeps only those keys; [] merges
    everything into one series per service."""
    if group_by is None:
        keep = attributes
    else:
        keep = {k: attributes[k] for k in group_by if k in attributes}
    return json.dumps(
        {"service": service_name, "attrs": keep}, sort_keys=True, default=str
    )


def counter_increment(prev: float | None, cur: float) -> float:
    "cur - prev, or cur after a reset (cur < prev) or when there is no prev."
    if prev is None or cur < prev:
        return cur
    return cur - prev


def rate(points: list[tuple[int, float]]) -> list[tuple[int, float | None]]:
    """[(time_ns, cumulative_value)] sorted by time -> [(time_ns, per_second)].
    First point has no rate (None)."""
    out: list[tuple[int, float | None]] = []
    prev_t = prev_v = None
    for t, v in points:
        if prev_t is None:
            out.append((t, None))
        else:
            dt = (t - prev_t) / NS
            out.append((t, counter_increment(prev_v, v) / dt if dt > 0 else None))
        prev_t, prev_v = t, v
    return out


def histogram_delta(prev: dict | None, cur: dict) -> dict:
    """Per-interval histogram from two cumulative points with identical
    explicit_bounds. Counts use counter_increment element-wise; sum likewise.
    prev=None (first point, or a reset detected on `count`) returns cur as-is."""
    if (
        prev is None
        or cur["count"] < prev["count"]
        or prev["explicit_bounds"] != cur["explicit_bounds"]
    ):
        return {**cur}
    return {
        **cur,
        "count": cur["count"] - prev["count"],
        "sum": None
        if cur.get("sum") is None or prev.get("sum") is None
        else cur["sum"] - prev["sum"],
        "bucket_counts": [
            c - p for c, p in zip(cur["bucket_counts"], prev["bucket_counts"])
        ],
        "min": None,  # min/max are not differentiable across intervals
        "max": None,
    }


def merge_histograms(points: Iterable[dict]) -> dict | None:
    "Element-wise sum of per-interval histograms sharing explicit_bounds."
    acc: dict | None = None
    for p in points:
        if acc is None:
            acc = {**p, "bucket_counts": list(p["bucket_counts"])}
            continue
        if acc["explicit_bounds"] != p["explicit_bounds"]:
            raise ValueError("mixed explicit_bounds in one series")
        acc["count"] += p["count"]
        acc["sum"] = (
            None
            if acc.get("sum") is None or p.get("sum") is None
            else acc["sum"] + p["sum"]
        )
        acc["bucket_counts"] = [
            a + b for a, b in zip(acc["bucket_counts"], p["bucket_counts"])
        ]
    return acc


def histogram_percentile(
    bounds: list[float],
    counts: list[int],
    q: float,
    minimum: float | None = None,
    maximum: float | None = None,
) -> float | None:
    """Estimate the q-quantile (0 < q <= 1) from explicit buckets.
    len(counts) == len(bounds) + 1; bucket i covers (bounds[i-1], bounds[i]],
    the last bucket is (bounds[-1], +inf). Linear interpolation within the
    bucket that contains rank q*total (Prometheus histogram_quantile rule)."""
    total = sum(counts)
    if total == 0 or not bounds:
        return None
    target = q * total
    cum = 0
    for i, c in enumerate(counts):
        prev_cum = cum
        cum += c
        if cum >= target and c > 0:
            if i == len(bounds):  # overflow bucket
                return maximum if maximum is not None else bounds[-1]
            upper = bounds[i]
            lower = (
                bounds[i - 1]
                if i > 0
                else (minimum if minimum is not None and minimum > 0 else 0.0)
            )
            return lower + (upper - lower) * ((target - prev_cum) / c)
    return maximum if maximum is not None else bounds[-1]


def last_per_bucket(points: list[dict], step_ns: int) -> dict[int, dict]:
    "points sorted by time_ns -> {bucket_start_ns: last point in that bucket}."
    out: dict[int, dict] = {}
    for p in points:
        out[bucket_start(p["time_ns"], step_ns)] = p
    return out


def aggregate_numbers(values: list[float], how: str) -> float:
    "how: 'sum' | 'avg' | 'last' | 'max'."
    if how == "sum":
        return sum(values)
    if how == "avg":
        return sum(values) / len(values)
    if how == "max":
        return max(values)
    return values[-1]

File: frontend/src/lib/metricsMath.ts
/** Pure metric math for the metrics pages. Mirrors
 * datasette_otel_viewer/metrics_math.py (same test vectors) — keep in sync. */

export const NS = 1_000_000_000;

/** Time-range presets shown by the picker, in seconds. */
export const RANGE_PRESETS: { label: string; seconds: number }[] = [
  { label: "15m", seconds: 15 * 60 },
  { label: "1h", seconds: 3600 },
  { label: "6h", seconds: 6 * 3600 },
  { label: "24h", seconds: 24 * 3600 },
  { label: "72h", seconds: 72 * 3600 },
];

const STEP_CHOICES = [5, 15, 30, 60, 300, 900, 1800, 3600];

/** Bucket width for a range: the smallest step giving <= ~180 buckets. */
export function stepForRange(rangeSeconds: number, maxBuckets = 180): number {
  for (const s of STEP_CHOICES) {
    if (rangeSeconds / s <= maxBuckets) return s;
  }
  return STEP_CHOICES[STEP_CHOICES.length - 1]!;
}

/** since/until in ns for a preset, from a ms clock (Date.now() in prod,
 * a fixed value in tests). until is floored to the step so the newest
 * bucket is complete-ish and stable between refreshes. */
export function rangeBounds(
  rangeSeconds: number,
  nowMs: number,
): { since_ns: number; until_ns: number; step_s: number } {
  const step = stepForRange(rangeSeconds);
  const stepNs = BigInt(step) * BigInt(NS);
  const untilNs =
    ((BigInt(Math.floor(nowMs)) * 1_000_000n) / stepNs) * stepNs + stepNs;
  const sinceNs = untilNs - BigInt(rangeSeconds) * BigInt(NS);
  return { since_ns: Number(sinceNs), until_ns: Number(untilNs), step_s: step };
}

/** Start of the bucket containing timeNs, given a bucket width in ns. */
export function bucketStart(timeNs: number, stepNs: number): number {
  return Math.floor(timeNs / stepNs) * stepNs;
}

/** cur - prev, or cur after a reset (cur < prev) / with no prev. */
export function counterIncrement(prev: number | null, cur: number): number {
  return prev === null || cur < prev ? cur : cur - prev;
}

/** Per-second rate from cumulative (t ns, value) pairs sorted by t. First
 * point → null. Zero dt → null. */
export function rate(
  points: { t: number; value: number }[],
): { t: number; value: number | null }[] {
  const out: { t: number; value: number | null }[] = [];
  let prev: { t: number; value: number } | null = null;
  for (const p of points) {
    if (prev === null) out.push({ t: p.t, value: null });
    else {
      const dt = (p.t - prev.t) / NS;
      out.push({
        t: p.t,
        value: dt > 0 ? counterIncrement(prev.value, p.value) / dt : null,
      });
    }
    prev = p;
  }
  return out;
}

/** Fill the [since, until) grid with nulls where a series has no bucket, so
 * the line breaks instead of interpolating across gaps. */
export function fillGaps<T extends { t: number }>(
  points: T[],
  sinceNs: number,
  untilNs: number,
  stepS: number,
): (T | { t: number; value: null })[] {
  const stepNs = stepS * NS;
  const byT = new Map(points.map((p) => [p.t, p]));
  const out: (T | { t: number; value: null })[] = [];
  for (let t = bucketStart(sinceNs, stepNs); t < untilNs; t += stepNs) {
    out.push(byT.get(t) ?? { t, value: null });
  }
  return out;
}

/** Prometheus histogram_quantile rule over explicit buckets;
 * counts.length === bounds.length + 1 (the last count is the overflow
 * bucket, > bounds[bounds.length - 1]). Linear interpolation within the
 * bucket that crosses the target rank; `min`/`max` (when known, e.g. from
 * an exemplar or a fixed instrumentation range) refine the open-ended
 * first/last buckets instead of assuming the bound itself. */
export function histogramPercentile(
  bounds: number[],
  counts: number[],
  q: number,
  min: number | null = null,
  max: number | null = null,
): number | null {
  const total = counts.reduce((a, b) => a + b, 0);
  if (total === 0 || bounds.length === 0) return null;
  const target = q * total;
  let cum = 0;
  for (let i = 0; i < counts.length; i++) {
    const c = counts[i]!;
    const prevCum = cum;
    cum += c;
    if (cum >= target && c > 0) {
      if (i === bounds.length) return max ?? bounds[bounds.length - 1]!;
      const upper = bounds[i]!;
      const lower = i > 0 ? bounds[i - 1]! : min !== null && min > 0 ? min : 0;
      return lower + (upper - lower) * ((target - prevCum) / c);
    }
  }
  return max ?? bounds[bounds.length - 1]!;
}
